zokrates_core/src/static_analysis/propagate_unroll.rs

```diff
@@ -0,0 +1,209 @@
+//! Module containing iterative unrolling in order to unroll nested loops with variable bounds
+//!
+//! For example:
+//! ```zokrates
+//! for field i in 0..5 do
+//! 	for field j in i..5
+//! 		//
+//! 	endfor
+//! enfor
+//! ```
+//!
+//! We can unroll the outer loop, but to unroll the inner one we need to propagate the value of `i` to the lower bound of the loop
+//!
+//! This module does exactly that:
+//! - unroll the outter loop, detecting that it cannot unroll the inner one as the lower `i` bound isn't constant
+//! - apply constant propagation to the program, *not visiting statements of loops whose bounds are not constant yet*
+//! - unroll again, this time the 5 inner loops all have constant bounds
+//!
+//! In the case that a loop bound cannot be reduced to a constant, we rely on a maximum number of passes after which we conclude that a bound is not constant
+//! This sets a hard limit on the number of loops with variable bounds in the program
+//!
+//! @file propagation.rs
+//! @author Thibaut Schaeffer <thibaut@schaeff.fr>
+//! @date 2018
+
+use static_analysis::propagation::Propagator;
+use static_analysis::unroll::Unroller;
+use typed_absy::TypedProgram;
+use zokrates_field::field::Field;
+
+pub struct PropagatedUnroller;
+
+const MAX_DEPTH: usize = 100;
```

Comment:
This seems a bit risky. Why is this sufficient?

Reply:
If we define a more or less arbitrary upper bound, it may be worth having a testcase that actually runs into this case.